src/hardware/mouse/mouse_manymouse.cpp

```diff
@@ -0,0 +1,481 @@
+		ManyMouse_Quit();
+		malfunction = true;
+		return;
+	}
+	initialized = true;
+
+	if (num_mice >= max_mice) {
+		num_mice           = max_mice - 1;
+		static bool logged = false;
+		if (!logged) {
+			logged = true;
+			LOG_ERR("MOUSE: Up to %d simultaneously connected mice supported",
+			        max_mice);
+		}
+	}
+
+	const auto new_driver_name = std::string(ManyMouse_DriverName());
+	if (new_driver_name != driver_name) {
+		driver_name = new_driver_name;
+		LOG_INFO("MOUSE: ManyMouse driver '%s'", driver_name.c_str());
+	}
+
+	Rescan();
+}
+
+void ManyMouseGlue::ShutdownIfSafe()
+{
+	if (is_mapping_in_effect || config_api_counter)
+		return;
+
+	ShutdownForced();
+}
+
+void ManyMouseGlue::ShutdownForced()
+{
+	if (!initialized)
+		return;
+
+	PIC_RemoveEvents(manymouse_tick);
+	ManyMouse_Quit();
+	ClearPhysicalMice();
+	num_mice    = 0;
+	initialized = false;
+}
+
+void ManyMouseGlue::StartConfigAPI()
+{
+	// Config API object is being created
+	++config_api_counter;
+	assert(config_api_counter > 0);
+}
+
+void ManyMouseGlue::StopConfigAPI()
+{
+	// Config API object is being destroyed
+	assert(config_api_counter > 0);
+	config_api_counter--;
+	ShutdownIfSafe();
+	if (!config_api_counter)
+		rescan_blocked_config = false;
+}
+
+void ManyMouseGlue::ClearPhysicalMice()
+{
+	mouse_info.physical.clear();
+	physical_devices.clear();
+	rel_x.clear();
+	rel_y.clear();
+}
+
+void ManyMouseGlue::Rescan()
+{
+	if (config_api_counter)
+		// Do not allow another rescan until MouseConfigAPI
+		// stops being used, it would be unsafe due to possible
+		// changes of physical device list size/indices
+		rescan_blocked_config = true;
+
+	ClearPhysicalMice();
+
+	for (uint8_t idx = 0; idx < num_mice; idx++) {
+		const auto name_utf8 = ManyMouse_DeviceName(idx);
+		std::string name;
+		UTF8_RenderForDos(name_utf8, name);
+
+		const char character_nbsp  = 0x7f; // non-breaking space
+		const char character_space = 0x20;
+
+		for (auto pos = name.size(); pos > 0; pos--) {
+			// Replace non-breaking space with a regular space
+			if (name[pos - 1] == character_nbsp)
+				name[pos - 1] = character_space;
+			// Remove non-ASCII and control characters
+			if (name[pos - 1] < character_space ||
+			    name[pos - 1] >= character_nbsp)
+				name.erase(pos - 1, 1);
+		}
+
+		// Try to rework into something useful names in the forms
+		// 'FooBar Corp FooBar Corp Incredible Mouse'
+		size_t pos = name.size() / 2 + 1;
+		while (--pos > 2) {
+			if (name[pos - 1] != ' ')
+				continue;
+			const std::string tmp = name.substr(0, pos);
+			if (name.rfind(tmp + tmp, 0) == std::string::npos)
+				continue;
+			name = name.substr(pos);
+			break;
+		}
+
+		// ManyMouse should limit device names to 64 characters,
+		// but make sure name is indeed limited in length, and
+		// strip trailing spaces
+		name.resize(std::min(static_cast<size_t>(64), name.size()));
+		while (!name.empty() && name.back() == ' ')
+			name.pop_back();
+
```

Comment:
This string processing is pretty involved and not too easy to follow (position division and addition, then substr'ing on position, etc).

Maybe just some basic cleanup - like `trim(name)` and `strip_punctuation(name)` is sufficient for the user to understand their mouse name?

```

void trim(std::string &str, const char trim_chars[] = " \r\t\f\n");
void strip_punctuation(std::string &str);

```

in `string_utils.h`.

Reply:
(or if your routine does genel string sanitization, consider moving it to `string_utils.h/cpp`.)

Reply:
The `trim` was useful. As for punctuations, we want to keep them. Used `std::replace` to simplify the code a bit, improved comments. Unfortunately, not everything could be simplified.

The name cleanup is specific to mouse names, unfortunately.